### PR TITLE
Game logic

### PR DESCRIPTION
**Pull request by Ashah15**

### CODE REVIEW

File: lib/game_logic.rb [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+class Player
+    public
+    def validate_rol
+        get_rol
+        if @rol === 1 || @rol === 2
+            board = Board.new ["1","2","3","4","5","6","7","8","9"], @rol, @name_1, @name_2
+            board.print_board
+            
+        else
+            puts "\n\n\nNumber #{@rol} is not a valid option"
```
Comment (maelfosso) [inline lib/game_logic.rb]: No puts or get.chomps here, please

File: lib/game_logic.rb [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+class Rules < Player
+    def winner?
+        #Horizontal Test
+        
+        if @board[0] === @draw && @board[1] === @draw && @board[2] === @draw
+            @board[0] = "-" 
+            @board[1] = "-" 
+            @board[2] = "-"
+            print_last
+            puts "\nPlayer #{@name} Wins"
```
Comment (maelfosso) [inline lib/game_logic.rb]: Please, **puts** is not allowed in the logic

File: lib/game_logic.rb [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+            @board[6] = "/"
+            print_last
+            puts "\nPlayer #{@name} Wins"
+            answer = again?
+            if answer === true
+                validate_rol 
+            end
+        end
+        
+    end
```
Comment (maelfosso) [inline lib/game_logic.rb]: The `winner?` method is too long.
I suggest you break it down it 3.
Also, why not move the responsibility to check whether the board is in a winning state to the **Board** class? Don't you think that it's is responsability ?

File: lib/game_logic.rb [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        print_board
+        input_cell
+    end
+    
+    def valid_move? 
+    if @rol == 1
+        @draw = "X"
+    else
+        @draw = "O"
+    end
```
Comment (maelfosso) [inline lib/game_logic.rb:36]: Please, indent your code to make clearer.

File: lib/game_logic.rb [hunk 5]
```diff
@@ -0,0 +1,209 @@
+require_relative '../bin/main'
+
+include UserInteraction
```
Comment (maelfosso) [inline lib/game_logic.rb]: Sorry, but that's is not allowed

You can't import the user interface into the game logic.
You should do the inverse.

File: lib/game_logic.rb [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+       puts "END OF GAME!"
+        puts "\n #{@board[0]} | #{@board[1]} | #{@board[2]} "
+        puts "-----------"
+        puts " #{@board[3]} | #{@board[4]} | #{@board[5]} "
+        puts "-----------"
+        puts " #{@board[6]} | #{@board[7]} | #{@board[8]} " 
+    end
+end
+player = Player.new
+player.validate_rol
```
Comment (maelfosso) [inline lib/game_logic.rb]: The game must be launch in the user interface, from `bin/main.rb` as indicated in the project description.

Keep separate the **Front-end (main.rb)** of the project to the **Back-end (the logic)**

File: lib/game_logic.rb [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+#===============================================================================================================================================
+# This is the class Rules. This class contains all rules of the game, so everytime a player plays, this class will verify if movements are valid
+# This class also checks if there is a winner or in case the game ends it will puts Game Over
+#===============================================================================================================================================
+
+class Rules < Player
+    def winner?
+        #Horizontal Test
+        
+        if @board[0] === @draw && @board[1] === @draw && @board[2] === @draw
```
Comment (maelfosso) [inline lib/game_logic.rb]: where does `@board` come from? Same thing for `@draw`?
Where did you declare them? Initialize them?

File: lib/game_logic.rb [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+            @board[1] = "-" 
+            @board[2] = "-"
+            print_last
+            puts "\nPlayer #{@name} Wins"
+            again?
+        elsif @board[3] === @draw && @board[4] === @draw && @board[5] === @draw
+            @board[3] = "-" 
+            @board[4] = "-" 
+            @board[5] = "-"
+            print_last
```
Comment (maelfosso) [inline lib/game_logic.rb]: `print_last` is not defined in that class.
It's defined in `Board` class. You can't call it like this.

You must separated the class. I suggest you put them in their own file to see how you implementation is not working.

File: lib/game_logic.rb [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+            puts "\nPlayer #{@name} Wins"
+            again?
+        elsif @board[6] === @draw && @board[7] === @draw && @board[8] === @draw
+            @board[6] = "-" 
+            @board[7] = "-" 
+            @board[8] = "-"
+            print_last
+            puts "\nPlayer #{@name} Wins"
+            again?
```
Comment (maelfosso) [inline lib/game_logic.rb]: Where does `again?` come from ?

**Review by maelfosso — CHANGES_REQUESTED** [review]
Hello @Ashah15 @jucora 

Congratulations, I wasn't easy to reach to that milestone. :clap: 

Now, It seems like you miss something because I can't play the game. `ruby bin/main.rb` doesn't display anything.
Please, take a look to the comment below to know what changes must be done.

In summary,
- You should keep separate the user interface (**Front end**) to the logic(**Back end**). The back-end must serve the front-end
- **No puts or get.chomps** in the logic
- Puts method in the class which has the first responsibility. The way you organize the method into class should be improved.
-  **Separate the class**. Put each class into its own file.

Kindly make the corresponding changes and re-submit a new code review request.
Thanks

**Comment by jucora** [comment]
Hello, thanks for code review.

A lot of changes were made, please let us know what you think, thank you so much!

File: bin/main.rb [hunk 10]
```diff
@@ -1,44 +1,110 @@
...
+  player2 = main.setPlayer2(player1)
+  board = Board.new()
+  while true
+    break if main.option(player1, board, game)
+    break if main.option(player2, board, game)
+  end
+  if main.endGame(game, board)
+    game.play_again = true
+    board.plays = 0
+  end
```
Comment (Oghenebrume50) [inline bin/main.rb]: This logic should not be here, this is the game loop can we put it among the game logic? you could create a game start or game loop method and just call it here

File: lib/player.rb [hunk 11]
```diff
@@ -1,8 +1,9 @@
 class Player
```
Comment (Oghenebrume50) [inline lib/player.rb:1]: The player class should do more think about what a player does in the TIC-TAC-TOE game and put it here. 
Example: A player selects boxes

File: lib/game.rb [hunk 12]
```diff
@@ -0,0 +1,14 @@
+
+require_relative 'player.rb'
+require_relative 'board.rb'
+
+class Game
```
Comment (Oghenebrume50) [inline lib/game.rb]: This class should hold the main game logic. The game loop, game start should be written here. This is what OOP is about, you have the OOP arrangement but not the functionality

File: lib/board.rb [hunk 13]
```diff
@@ -0,0 +1,58 @@
+
+
+class Board
+  attr_accessor :winner, :plays
+  def initialize
+    @plays = 0
+    @wins1 = { 1 => [1, 2, 3], 2 => [4, 5, 6], 3 => [7, 8, 9], 4 => [1, 4, 7] }
+    @wins2 = { 5 => [2, 5, 8], 6 => [3, 6, 9], 7 => [1, 5, 9], 8 => [3, 5, 7] }
+    @wins = [@wins1, @wins2]
```
Comment (Oghenebrume50) [inline lib/board.rb]: Everything about winning should be written in the game class. Think about it, it is the game that determines who wins not the board :wink:

**Review by Oghenebrume50 — CHANGES_REQUESTED** [review]
Hi team, you have done a good job so far,  :+1:  but I have outlined some changes that needed to be done in order to make this OOP oriented if there are any misunderstandings please reach out to me on Slack. Happy coding guys :+1: :wink:

**Comment by jucora** [comment]
Hello.

Thanks for code review. Changes were done. Now the main section looks more clear and the game logic has the method that you asked.

Also two more methods were created in the logic_game section to validate when the game ends and also to validate when player selects an option to play.

The board class has a method to print the board every time a gamer plays.

Please let us know what you think


Thanks

File: lib/board.rb [hunk 14]
```diff
@@ -0,0 +1,24 @@
+require_relative 'game_logic.rb'
+class Board
+attr_accessor :map
+  def initialize
+    $map = []
```
Comment (Ispirett) [inline lib/board.rb]: [Optional]  Not required to complete this project  Making the variable  global  allows it to be accessible 
 from anywhere in your program. You would not want someone to change the turn of the game. Encapsulation should be used here.

File: bin/main.rb [hunk 15]
```diff
@@ -1,44 +1,95 @@
...
+      print "\n ***** #{@name_1.capitalize} Wins ***** \n"
+    elsif game.winner == 2
+      puts @print_board.display_board
+      print "\n ***** #{@name_2.capitalize} Wins ***** \n"
+    else
+      puts @print_board.display_board
+      puts "GAME OVER... No winners this time!"
+    end
+    
+    while @again == false do
```
Comment (Ispirett) [inline bin/main.rb:74]: This method has a lot going on. You should place this section in it's own method eg `replay_game`

**Review by Ispirett — CHANGES_REQUESTED** [review]
Hey @Ashah15 and @jucora Great work with the updates project is looking great.
image. 

You can merge this project :1st_place_medal:

File: lib/board.rb (hunk 14 above)
Comment (jucora) [reply lib/board.rb]: done :)

File: bin/main.rb (hunk 15 above)
Comment (jucora) [reply bin/main.rb:74]: done :)

**Pull request merged**